### ISSUE TITLE
Prevent composition from being overwritten

### ISSUE THREAD

**Issue by JSkimming** [JSkimming/abioc#14]
Except for where [SingleConstructorRegistrationVisitor](https://github.com/JSkimming/abioc/blob/v0.3.0/src/Abioc/Composition/Visitors/SingleConstructorRegistrationVisitor.cs#L43-L48) is checking.

This is how we can pick-up conflicting registrations, like a two factories registered for the same type.

**Comment by jzabroski** [comment]
Do you have a test case for this? I'd love to contribute and get my hands on Roslyn, and this seems like an easy way to contribute.

**Comment by JSkimming** [comment]
@jzabroski Thanks for the interest, I've not yet devised a test, I've been using the issues as a simple TODO list for myself.

Let me see see if I can knock up a quick test and get a broken build.

**Comment by jzabroski** [comment]
Thanks. Just saw the benchmarks and it blows my mind how in a few short weeks you took the #1 spot in Daniel's benchmarks.

**Comment by JSkimming** [comment]
@jzabroski Thanks, it's all down to generating the code, and relying on the Roslyn compiler optimising it further.

FYI, most of the tests output the generated code, see the [CI build](https://ci.appveyor.com/project/JSkimming/abioc/build/tests). Also, if you download the [TestResults.zip](https://ci.appveyor.com/project/JSkimming/abioc/build/artifacts) artifact, you can find the xUnit generated test results HTML file, as wall as all the code coverage results. I'm trying to maintain high code coverage.

Just a note of caution, I'm sticking with a `0.n` version for a little while until I've implemented a greater feature set, and it's bedded in within the live system I'm currently using it. As a result, the API may include breaking changes. Once I get to 1.0 (I don't plan on taking long) I'll follow [semantic versioning](http://semver.org/), and commit to maintain backwards compatibility within major versions.

**Comment by JSkimming** [comment]
@jzabroski I've just merged #29 that introduces a bunch of tests, several are tests that pass that should continue to pass, but there are three [like this](https://github.com/JSkimming/abioc/blob/1c48d3a148abb6669e41402c33fc23d39f9a4799/test/Abioc.Tests/PreventCompositionOverwriteTests.cs#L259-L274), that are skipped because they fail but they express the desired functionality.

I was thinking this could be fixed by changing [CompositionContext](https://github.com/JSkimming/abioc/blob/v0.3.0/src/Abioc/Composition/CompositionContext.cs#L29) to expose the Compositions as a `IReadOnlyDictionary` with a backing field (that's still a `Dictionary`).

```csharp
public IReadOnlyDictionary<Type, IComposition> Compositions => _compositions;
```

Then adding an `AddComposition` method, that throws an exception should any composition be overwritten that is not a [SingleConstructorRegistrationVisitor](https://github.com/JSkimming/abioc/blob/v0.3.0/src/Abioc/Composition/Visitors/SingleConstructorRegistrationVisitor.cs).

This would allow the change to be in a single place (SRP), and the encapsulation feels better (I never quite liked the raw dictionary exposed like that).
